hopper/mainloop_fwd_sm90_tma_gmma_ws.hpp

```diff
@@ -454,25 +540,41 @@ struct CollectiveMainloopFwd {
         // for fp8, change from NumThreadsPerWarp to NumThreadsPerWarpGroup
         cutlass::arch::NamedBarrier::sync(NumMmaThreads + cutlass::NumThreadsPerWarpGroup, static_cast<int>(FwdNamedBarriers::QueryEmpty) /*id*/);
 
-        if constexpr(Is_causal) {
+        if (warp_idx_in_warpgroup == 0 && lane_predicate) {
+            shared_storage.barrier_Q.arrive_and_expect_tx(TmaTransactionBytesQ);
+            copy(mainloop_params.tma_load_Q.with(reinterpret_cast<cutlass::arch::ClusterTransactionBarrier::ValueType&>(shared_storage.barrier_Q), 0 /*mcast_mask*/), tQgQ, tQsQ);
+            if constexpr(!Ktraits::VO_union_all) {
+                pipeline_v.producer_acquire(smem_pipe_write);
+                copy(mainloop_params.tma_load_V.with(*pipeline_v.producer_get_barrier(smem_pipe_write), mcast_mask_kv),
+                    tVgV(_, n_block), tVsV(_, smem_pipe_write.index()));
+            }
+
+        }
+        // With fp8 kernel, smem_o is in union with smem_v_out,
+        // except for split kernel + hdim 256,
+        // so could use NamedBarrier instead of ClusterBarrier.
+        // But, this doesn't appear to have any benefit.
+        shared_storage.barrier_O.wait((work_idx + 1) % 2);
+
+        if constexpr(Ktraits::VO_union_all) {
             if (warp_idx_in_warpgroup == 0 && lane_predicate) {
-                shared_storage.barrier_Q.arrive_and_expect_tx(TmaTransactionBytesQ);
-                copy(mainloop_params.tma_load_Q.with(reinterpret_cast<cutlass::arch::ClusterTransactionBarrier::ValueType&>(shared_storage.barrier_Q), 0 /*mcast_mask*/), tQgQ, tQsQ);
                 pipeline_v.producer_acquire(smem_pipe_write);
                 copy(mainloop_params.tma_load_V.with(*pipeline_v.producer_get_barrier(smem_pipe_write), mcast_mask_kv),
                     tVgV(_, n_block), tVsV(_, smem_pipe_write.index()));
             }
+        }
 
-            shared_storage.barrier_O.wait((work_idx + 1) % 2);            
-                        
+        // if constexpr(Is_causal || Is_local || Is_split) {
+        {
+            #if 0
```

Comment:
Why not just removing deprecated code?

Reply:
Removed it -- I thought it was convenient to leave in when testing fp8 kernel perf, but it does clutter the code more.